packages/react-instantsearch/src/ui/InternalHighlight.tsx

```diff
@@ -1,101 +1,9 @@
-import React, { Fragment } from 'react';
+import { createHighlightComponent } from 'instantsearch-ui-components';
+import { createElement, Fragment } from 'react';
 
-import { cx } from './lib/cx';
+import type { Pragma } from 'instantsearch-ui-components';
 
-type HighlightPartProps = {
-  children: React.ReactNode;
-  classNames: InternalHighlightClassNames;
-  highlightedTagName: React.ElementType;
-  nonHighlightedTagName: React.ElementType;
-  isHighlighted: boolean;
-};
-
-function HighlightPart({
-  classNames,
-  children,
-  highlightedTagName,
-  isHighlighted,
-  nonHighlightedTagName,
-}: HighlightPartProps) {
-  const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;
-
-  return (
-    <TagName
-      className={
-        isHighlighted ? classNames.highlighted : classNames.nonHighlighted
-      }
-    >
-      {children}
-    </TagName>
-  );
-}
-
-type HighlightedPart = {
-  isHighlighted: boolean;
-  value: string;
-};
-
-export type InternalHighlightClassNames = {
-  /**
-   * Class names to apply to the root element
-   */
-  root: string;
-  /**
-   * Class names to apply to the highlighted parts
-   */
-  highlighted: string;
-  /**
-   * Class names to apply to the non-highlighted parts
-   */
-  nonHighlighted: string;
-  /**
-   * Class names to apply to the separator between highlighted parts
-   */
-  separator: string;
-};
-
-export type InternalHighlightProps = React.HTMLAttributes<HTMLSpanElement> & {
-  classNames: InternalHighlightClassNames;
-  highlightedTagName?: React.ElementType;
-  nonHighlightedTagName?: React.ElementType;
-  separator?: React.ReactNode;
-  parts: HighlightedPart[][];
-};
-
-export function InternalHighlight({
-  parts,
-  highlightedTagName = 'mark',
-  nonHighlightedTagName = 'span',
-  separator = ', ',
-  className,
-  classNames,
-  ...props
-}: InternalHighlightProps) {
-  return (
-    <span {...props} className={cx(classNames.root, className)}>
-      {parts.map((part, partIndex) => {
-        const isLastPart = partIndex === parts.length - 1;
-
-        return (
-          <Fragment key={partIndex}>
-            {part.map((subPart, subPartIndex) => (
-              <HighlightPart
-                key={subPartIndex}
-                classNames={classNames}
-                highlightedTagName={highlightedTagName}
-                nonHighlightedTagName={nonHighlightedTagName}
-                isHighlighted={subPart.isHighlighted}
-              >
-                {subPart.value}
-              </HighlightPart>
-            ))}
-
-            {!isLastPart && (
-              <span className={classNames.separator}>{separator}</span>
-            )}
-          </Fragment>
-        );
-      })}
-    </span>
-  );
-}
+export const InternalHighlight = createHighlightComponent({
+  createElement: createElement as Pragma,
```

Comment:
does createElement not match Pragma? is Pragma typed wrong?

Reply:
It's a complex topic, because Preact and React types aren't fully compatible. I picked the types from `@algolia/ui-components` which were initially meant to be as generic as possible, but we never really spent time making sure they are.

IMO that's too big of a topic to be settled in this PR. I'd love to spend some time reinforcing our UI types so they're really generic, but type casting could do the job until we have time to spend on it, as there's sufficient overlap for TypeScript to be happy.

Reply:
LGTM, in a future version we can indeed make Pragma more compatible / generic